File: 2872-maximum-number-of-k-divisible-components.py
from collections import defaultdict

class Solution:
    def maxKDivisibleComponents(self, n: int, edges: list[list[int]], values: list[int], k: int) -> int:
        '''
        Treat this as a tree (graph without cycles).

        Key insight: Any subtree with a sum divisible by k can be split off 
        as a separate component. We count all such possible components by:
            1. Building adjacency list for the tree
            2. Using DFS to calculate subtree sums bottom-up
            3. Whenever a subtree sum is divisible by k, increment the count
            4. Track parent to avoid revisiting nodes in the tree traversal
        '''
        adj_list = defaultdict(list)
        for first, second in edges:
            adj_list[first].append(second)
            adj_list[second].append(first)

        res = 0
        def dfs(cur, parent):
            total = values[cur]

            for child in adj_list[cur]:
                if child != parent:
                    total += dfs(child, cur)

            if total % k == 0:
                nonlocal res
                res += 1

            return total

        dfs(0, -1)
        return res
        
        
sol = Solution()
print(f'output: {sol.maxKDivisibleComponents(n = 5, edges = [[0,2],[1,2],[1,3],[2,4]], values = [1,8,1,4,4], k = 6)}, expected: 2')
print(f'output: {sol.maxKDivisibleComponents(n = 7, edges = [[0,1],[0,2],[1,3],[1,4],[2,5],[2,6]], values = [3,0,6,1,5,2,1], k = 3)}, expected: 3')

File: 2872-maximum-number-of-k-divisible-components.java
import java.util.*;

class MaximumNumberOfKDivisibleComponents {

    public int maxKDivisibleComponents(int n, int[][] edges, int[] values, int k) {
        Map<Integer, List<Integer>> adjList = new HashMap<>();
        for (int[] edge : edges) {
            int first = edge[0], second = edge[1];
            List<Integer> firstChildren = adjList.getOrDefault(first, new ArrayList<>());
            firstChildren.add(second);
            adjList.put(first, firstChildren);
            List<Integer> secondChildren = adjList.getOrDefault(second, new ArrayList<>());
            secondChildren.add(first);
            adjList.put(second, secondChildren);
        }
        int[] res = new int[1];
        dfs(adjList, values, k, 0, -1, res);

        return res[0];
    }

    private long dfs(Map<Integer, List<Integer>> adjList, int[] values, int k, int cur, int parent, int[] res) {
        long total = values[cur];
        for (int child : adjList.getOrDefault(cur, new ArrayList<>())) {
            if (child != parent) {
                total += dfs(adjList, values, k, child, cur, res);
            }
        }
        if (total % k == 0) {
            res[0]++;
        }
        return total;
    }

    public static void main(String[] args) {
        MaximumNumberOfKDivisibleComponents sol = new MaximumNumberOfKDivisibleComponents();
        System.out.println(
                "Output: " + sol.maxKDivisibleComponents(5, new int[][] { { 0, 2 }, { 1, 2 }, { 1, 3 }, { 2, 4 } },
                        new int[] { 1, 8, 1, 4, 4 }, 6) + ", expected: 2");
        System.out.println("Output: " + sol.maxKDivisibleComponents(7,
                new int[][] { { 0, 1 }, { 0, 2 }, { 1, 3 }, { 1, 4 }, { 2, 5 }, { 2, 6 } },
                new int[] { 3, 0, 6, 1, 5, 2, 1 }, 3) + ", expected: 3");
    }
}
